### PR TITLE
WMDP-210: Create a generic template library for a flask API

### PR DESCRIPTION
**Pull request by chouinar**
## Ticket

https://wicmtdp.atlassian.net/browse/WMDP-210
https://github.com/navapbc/template-application-flask/issues/3

## Changes
Modifies the WIC MT Demo application to be a more generic implementation, but retain all the features implemented (except some unfinished infra/CI pieces that we can port over later).

The current implementation only aims to be runnable locally, but some upcoming work is implementing the necessary configurations to get this running in AWS.

## Context for reviewers
This PR only showcases the diff with the most recent changes of the WIC MT app (ie. just the work to make it more generic), and doesn't highlight all of the components that were built. To see those, you'll probably want to check this branch out and look around, but a few notes:
* API with a post endpoint that writes to the DB using basic apikey authentication.
* Postgres DB + alembic migrations configured for updating the DB when you update the sqlalchemy DB models.
* Thorough formatting & linting tools.
* Logging, with formatting in both human-readable and JSON formats.
* Ability to write backend scripts connected to the DB with an example one that generates a CSV (locally or on S3 with proper credentials).
* Ability to run the various utility scripts inside or outside of Docker (Note: I may need someone to test the outside as it requires installing a few libraries and/or configuring a few things that I need to document).
* Restructured and improved API request and response error handling which gives more details than the out-of-the-box approach for both Connexion and Pydantic
* Sane environment variable configuration for local development (`local.env`)

This PR is mostly focused on getting things more generic, but if you see any updates you'd like to the existing features, please add a note and we can create an issue to come back to that in a follow-up.

## Testing

`make clean-volumes init run` will cleanup any past versions of this code you've checked out and get the DB + API running. You can verify the API is running by doing `make run-logs`. If you see something like this for the last few logs, you should be good to go:
Screen Shot 2022-09-13 at 3 04 40 PM

Going to `http://localhost:8080/v1/docs/` should bring up the swagger UI which will show the two currently implemented endpoints.
Screen Shot 2022-09-13 at 3 07 33 PM

Clicking authorize in the top-right and entering `LOCAL_AUTH_12345678` as the API key will authenticate you for calling the `POST /example-person` API. Using the default parameters, you can submit and get a response back:
Screen Shot 2022-09-13 at 3 14 40 PM

In addition, you should be able to query your local DB (here I'm using TablePlus to connect) and see:
Screen Shot 2022-09-13 at 3 16 57 PM

If you submit a few more records to the DB and then run `make generate-pet-csv` you can run the backend script to make a CSV with some of the info submitted to the API:
Screen Shot 2022-09-13 at 3 40 44 PM

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ## Ticket
 
-https://wicmtdp.atlassian.net/browse/{TICKET_ID}
+{TICKET_LINK}
```
Comment (chouinar) [inline .github/pull_request_template.md:3]: I probably want to make some obvious `REPLACE_ME` in a few places in the template.

File: app/api/db/migrations/versions/2022_09_12_16_18_41_64cca9571cf0_first_migration.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+revision = "64cca9571cf0"
+down_revision = None
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table(
+        "example_person",
```
Comment (lorenyu) [inline app/api/db/migrations/versions/2022_09_12_16_18_41_64cca9571cf0_first_migration.py:22]: nice i like that we're including example tables. on this note, could we consider creating more realistic tables that projects might typically encounter? for example, i think most projects would have a `user` table with an `id UUID` column, `created_at timestamp` column, and a `username TEXT` column.

not sure what would be a good second example table. some ideas could be:
```
application(id uuid, created_at timestamp, submitted_at timestamp)
eligibility_determination(determination_date date, determination text)
document(id uuid, url text)
```

File: app/api/db/models/__init__.py [hunk 3]
```diff
@@ -1 +1 @@
-from . import eligibility_models  # noqa: F401
+from . import example_models  # noqa: F401
```
Comment (lorenyu) [inline app/api/db/models/__init__.py]: nit: don't need to ignore F401, the correct way to handle this is to define `__all__`
```suggestion
from . import example_models

__all__ = "example_models"
```

File: app/api/db/models/factories.py [hunk 4]
```diff
@@ -57,19 +58,25 @@ class Meta:
         sqlalchemy_session_persistence = "commit"
```
Comment (lorenyu) [inline app/api/db/models/factories.py:58]: i googled `sqlalchemymodelfactory` but can't find anything on that. is that part of sqlalchemy, a library, or something we made?

File: app/api/db/models/factories.py [hunk 5]
```diff
@@ -57,19 +58,25 @@ class Meta:
...
 
 
-class EligibilityScreenerFactory(BaseFactory):
+class ExamplePersonFactory(BaseFactory):
```
Comment (lorenyu) [inline app/api/db/models/factories.py:61]: some high level thoughts
1. i think there's a simpler way to generate synthetic data. rather than subclassing SQLAlchemyModelFactory, we should just create a custom Faker provider. here's [an example in the CMS FHIR bid](https://github.com/navapbc/cms-ab2d-bcda-and-dpc/blob/main/api/test/data_generator/fhir_provider.py).
2. synthetic data generation could be test dependency but is ideally not a prod dependency (e.g. Faker should be a dev dependency), so let's move the fake data modules to the `test` folder

File: app/api/db/models/example_models.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
+from datetime import date
```
Comment (lorenyu) [inline app/api/db/models/example_models.py]: question for discussion, should we aim to have one model per file like `models/person.py` which exports `Person`, `models/user.py` which exports `User`, or a single module like `models/example_models.py` which has `User` and `Person` in the file? i'm biased towards having each file focus on one thing, but am interested in what others think.

question for my understanding: if we put multiple models in one file, then in practice what would you call that file i.e. what would `example_models.py` be called? `models/models.py` seems redundant.

File: app/api/route/handler/example_person_handler.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
+from datetime import date
+from typing import Optional
+from uuid import uuid4
+
+from pydantic import UUID4
+
+import api.logging
+from api.db.models.example_models import ExamplePerson, ExamplePet
+from api.util.api_context import ApiContext
+from api.util.pydantic_util import PydanticBaseModel
```
Comment (lorenyu) [inline app/api/route/handler/example_person_handler.py]: the problem with "util" modules is that they become an overly centralized dependency. this is a situation where DRY is actually harmful. instead, we should have something like `class BaseRequestModel` which we define in the `app.api.route` module or `app.api.handler` module, and if we use Pydantic in another context that isn't request handling (for example if our internal application data models layer also uses Pydantic), we should define a separate base class for that use case. The various contexts of the application should hide the underlying libraries from each other.

relatedly, in the context of request handling, I don't think we need to set `orm_mode = true`, since request_body is always a dictionary.

File: app/api/route/handler/example_person_handler.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    first_name: str
+    middle_name: Optional[str]
+    last_name: str
+    phone_number: str
+    date_of_birth: date
+    is_real: bool
+    pets: Optional[list[ExamplePetParams]]
+
+
+class ExamplePersonResponse(ExamplePersonParams):
```
Comment (lorenyu) [inline app/api/route/handler/example_person_handler.py:32]: for future discussion: i'm not sure i see much value in doing pydantic model validation on the server response.

**Review by lorenyu — COMMENTED** [review]
awesome to see!

A topic for discussion is whether we should include ORM as part of the template. I'm biased against using ORMs, i think the convenience isn't worth the added complexity. The extra work of creating repositories (see [repository pattern](https://martinfowler.com/eaaCatalog/repository.html)), or what I called "persistence layer" in the [server-side guidelines presentation i gave to mass pfml](https://docs.google.com/presentation/d/1QRgqRitXQ0wmSQB1CZZfN4m1eFyjF301JNae3USlFZA/edit#slide=id.p), isn't that big a deal, and makes very explicit what are all the queries the application makes, making it easier to design database indexes and plan database migrations.

File: app/api/db/models/example_models.py (hunk 6 above)
Comment (rocketnova) [reply app/api/db/models/example_models.py]: I strongly prefer one model per file.

File: app/local.env [hunk 9]
```diff
@@ -36,6 +36,11 @@ POSTGRES_PASSWORD=secret123
 # value specified in the docker-compose file
 DB_HOST=localhost
 
+# When an error occurs with a SQL query,
+# whether or not to hide the parameters which
+# could contain sensitive information.
+HIDE_SQL_PARAMETER_LOGS=TRUE
+
```
Comment (rocketnova) [inline app/local.env]: This seems useful to include in the non-generic repo.

File: app/api/db/migrations/versions/2022_09_12_16_18_41_64cca9571cf0_first_migration.py (hunk 2 above)
Comment (chouinar) [reply app/api/db/migrations/versions/2022_09_12_16_18_41_64cca9571cf0_first_migration.py:22]: That's roughly what I was aiming for here. The `ExamplePerson` table has each of those types of fields. I actually deliberately avoided making a "real" table like users as I didn't want a name conflict if a developer wanted to first add a `User` table but our example one was already there (even if I just called it `ExampleUser`)

File: app/api/db/models/example_models.py (hunk 6 above)
Comment (chouinar) [reply app/api/db/models/example_models.py]: I actually prefer a 3rd option, logically grouped files. 1-per-file seems like a lot of files once a project grows large enough. A quick check of the Mass code shows we have ~200 tables. If I needed to look up something with say payments, I would have needed to open a dozen files to see all the relevant tables.

Lets say you have a `person` file, and a few related tables that would only ever be attached to a person like lookup tables for race/gender/etc. Definitely would want it more split up than the Mass code (the employee model file in particular).

**Comment by chouinar** [comment]
> awesome to see!
> 
> A topic for discussion is whether we should include ORM as part of the template. I'm biased against using ORMs, i think the convenience isn't worth the added complexity. The extra work of creating repositories (see [repository pattern](https://martinfowler.com/eaaCatalog/repository.html)), or what I called "persistence layer" in the [server-side guidelines presentation i gave to mass pfml](https://docs.google.com/presentation/d/1QRgqRitXQ0wmSQB1CZZfN4m1eFyjF301JNae3USlFZA/edit#slide=id.p), isn't that big a deal, and makes very explicit what are all the queries the application makes, making it easier to design database indexes and plan database migrations.

It might be due to prior poor experiences querying DBs with effectively raw SQL, but the ORM pattern in Mass was lightyears better than any DB interaction I'd had on any prior project. I think having the typed objects made development monumentally easier, BUT admittedly don't know what alternatives might exist. Do you have any examples of DB interfacing and/or how migrations would be managed?

File: app/api/route/handler/example_person_handler.py (hunk 8 above)
Comment (chouinar) [reply app/api/route/handler/example_person_handler.py:32]: I definitely think the API validation as a whole could use a bit of work. I feel like more of that should be done in openapi (or equivalent). Except for potential business rules that require DB lookups or complex algorithms, should ideally not need Pydantic like this.

File: app/api/route/handler/example_person_handler.py (hunk 7 above)
Comment (chouinar) [reply app/api/route/handler/example_person_handler.py]: I could restructure the files a bit. Ended up with more in the util folder than I'd like. Some truly generic pieces like the string & file utils might make sense to keep there, but you're right that api-specific components should probably go.

I believe `orm_mode = True` is for the reverse. Being able to convert the SqlAlchemy object into the response object.

File: app/api/route/handler/example_person_handler.py (hunk 8 above)
Comment (chouinar) [reply app/api/route/handler/example_person_handler.py:32]: But I also loathe OpenAPIs validation as it's immensely finnicky, hard to follow, and missing a lot of pretty basic things. See: how I had to structure the request/response models in it for this PR just to have the ID fields not be in the request, but be in the response

File: app/api/db/models/factories.py (hunk 5 above)
Comment (chouinar) [reply app/api/db/models/factories.py:61]: 1. I don't have access to that repo to see the example.
2. You're probably right, kept a similar pattern to what Mass did. I think a few non-local only scripts did rely on the factory, but shouldn't have.

File: app/api/db/models/example_models.py (hunk 6 above)
Comment (lorenyu) [reply app/api/db/models/example_models.py]: Refining my language a bit to use nomenclature from domain driven design, if we define entities and value objects separately, where entities are "Objects that have a distinct identity that runs through time and different representations" and value objects are "Objects that matter only as the combination of their attributes", then things like `person`, `user`, `application` are entities and things like enums/lookup table values and addresses are value objects, one suggestion is that we have one "entity" per file. Related classes for that entity such as any enums could be in the same file.

- https://martinfowler.com/bliki/EvansClassification.html
- https://enterprisecraftsmanship.com/posts/entity-vs-value-object-the-ultimate-list-of-differences/

File: app/api/db/models/factories.py (hunk 4 above)
Comment (chouinar) [reply app/api/db/models/factories.py:58]: A part of factoryboy
https://factoryboy.readthedocs.io/en/stable/orms.html#sqlalchemy

**Comment by lorenyu** [comment]
Yeah, totally agree with strongly typed objects, not suggesting we use dictionary objects throughout our application code. Just suggesting that we manually do the conversion between SQL and typed objects ourselves rather than automatically with an ORM. A couple of examples:
- [pull_request_store](https://github.com/navapbc/DevOpsMetrics/blob/main/stores/pull_requests_store.py) in the DevOpsMetrics codebase. I just defined a `DBPullRequest` model as a native python dataclass, but it could have easily been a pydantic model and the code would work the same but with runtime validation in addition to compile-type type checking. (note: the "store" is a repository in the repository pattern, but i chose to call it "stores" in that context since that entire codebase is about processing data from github so calling it repository could be easily confused with github repository.)
- [repositories in the CMS FHIR API bid](https://github.com/navapbc/cms-ab2d-bcda-and-dpc/tree/main/api/cms/api/repositories) a bunch of repositories. in the examples we probably should have defined the models in a separate folder.

some notes:
- the entity models could probably have defined in a separate folder rather than with the repositories.
- we didn't actually have to write much code beyond the queries, since [psycopg contains a lot of useful row factories](https://www.psycopg.org/psycopg3/docs/advanced/rows.html) for converting from cursor records to objects.

**Comment by chouinar** [comment]
> 

Yeah, if we use dataclasses + some utils on top of it, that would help with the object issues (Pydantic might also have some value, but probably more hassle than it's worth). How do you manage migrations in this approach?

Manually writing the SQL can be a bit more risky, are there any utilties/libraries/approaches for helping manage that (does alembic have a way to handle that?).

In either case, how should we proceed here? I don't want to try and rework this in this PR. Any major feature changes I'd like to be follow-ups to keep the scope from growing.

File: app/local.env (hunk 9 above)
Comment (chouinar) [reply app/local.env]: Added to the cleanup PR and merged that, so it's upstream of this change now.

File: .github/pull_request_template.md (hunk 1 above)
Comment (rocketnova) [reply .github/pull_request_template.md:3]: Good call

File: app/api/route/handler/example_person_handler.py (hunk 8 above)
Comment (rocketnova) [reply app/api/route/handler/example_person_handler.py:32]: I also think OpenAPI's validation is really finicky and hard to follow

File: app/api/db/models/example_models.py (hunk 6 above)
Comment (rocketnova) [reply app/api/db/models/example_models.py]: @lorenyu 
> Related classes for that entity such as any enums could be in the same value.

Do you mean in the same file?

Comment (rocketnova) [reply app/api/db/models/example_models.py]: @chouinar 
> I actually prefer a 3rd option, logically grouped files. 1-per-file seems like a lot of files once a project grows large enough. A quick check of the Mass code shows we have ~200 tables. If I needed to look up something with say payments, I would have needed to open a dozen files to see all the relevant tables.

Yah that's fair and I agree with that too.

Comment (chouinar) [reply app/api/db/models/example_models.py]: Might be better for me to think of it like "the big things" or the top-level objects. (And sidenote: it might just be the example used, but I find a lot of flaws with the ideas argued in that second link).

But I think this is getting a bit sidetracked. If I renamed the file to `example_person.py` would that make more sense? The person is the main entity, and pets are a subentity that goes with them.

File: app/api/route/handler/example_person_handler.py (hunk 8 above)
Comment (chouinar) [reply app/api/route/handler/example_person_handler.py:32]: Might be worth looking into alternatives. Openapis main benefit is the swagger page, which I'm not sure we'd get in another way.

We do already have the `CustomRequestBodyValidator` implemented, but only add one or two generic errors. Perhaps we could do something with that?

Comment (chouinar) [reply app/api/route/handler/example_person_handler.py:32]: Actually, I remember coming across it a few times, but it seems like FastAPI https://github.com/tiangolo/fastapi is "write pydantic models, generate swagger" which might be better.

Comment (chouinar) [reply app/api/route/handler/example_person_handler.py:32]: Yeah, it seems like Flask it pretty tightly coupled to openapi, you have to specify an openapi to even initialize an API. 

While there are some generators out there, they don't necessarily look easier to work with (mostly just adding annotations filled with JSON/yaml to the entry-point methods). The best chance to have the models defined in one place would be to pivot the API entirely to FastAPI which seems to have been built explicitly to work this way,  but that would probably be a bit of work as we'd need to rework all of the API code.

Comment (chouinar) [reply app/api/route/handler/example_person_handler.py:32]: (Maybe this? https://github.com/kuimono/openapi-schema-pydantic )

**Comment by lorenyu** [comment]
> In either case, how should we proceed here? I don't want to try and rework this in this PR. Any major feature changes I'd like to be follow-ups to keep the scope from growing.

Agree, sounds good to not block this PR on the ORM discussion.

> Yeah, if we use dataclasses + some utils on top of it, that would help with the object issues (Pydantic might also have some value, but probably more hassle than it's worth). How do you manage migrations in this approach?

I'm still open to using alembic for migrations, even though it depends on SQLAlchemy. We'd have to write our own migrations rather than using alembic's auto-generate migrations feature, but I also think that's reasonable. And even though we'd still have SQLAlchemy as a dependency, the application doesn't need to use it.

> Manually writing the SQL can be a bit more risky, are there any utilties/libraries/approaches for helping manage that (does alembic have a way to handle that?).

Which part of it is risky? Do you mean from a security perspective? I think as long as people aren't ever interpolating into queries we should be fine. Every sql library including psycopg itself will escape things for you as long as you use it correctly e.g.  do this `cursor.execute('SELECT * FROM user WHERE id = %s', id)` not this `cursor.execute(f'SELECT * FROM user WHERE id = {id})`

File: app/api/route/handler/example_person_handler.py (hunk 7 above)
Comment (lorenyu) [reply app/api/route/handler/example_person_handler.py]: > I believe orm_mode = True is for the reverse. Being able to convert the SqlAlchemy object into the response object.

yeah that makes sense, i guess we can defer this conversation until the future where we resolve the other [discussion on doing pydantic model validation on the server response](https://github.com/navapbc/wic-mt-demo-project-mock-api/pull/20#discussion_r970143239)

File: app/api/db/models/example_models.py (hunk 6 above)
Comment (lorenyu) [reply app/api/db/models/example_models.py]: if pet is a top level resource (e.g. we're going to have API routes for them) then i'd have a separate file for it, but if it's only ever relevant in the context of a person, then keeping it in the person file makes sense.

to my point about realistic examples, i don't think the concept of `pet` is close enough to our projects for it to provide a useful reference for when an engineer is determining whether or not to include something like `address` in the same file or not.

File: app/api/db/models/factories.py (hunk 5 above)
Comment (lorenyu) [reply app/api/db/models/factories.py:61]: 1. weird, i'll try to get you access, in the meantime feel free to consider this discussion non-blocking (can address in a future PR)

Comment (chouinar) [reply app/api/db/models/factories.py:61]: Moved the factory to the test folders + moved its dependencies to be dev only.

Comment (lorenyu) [reply app/api/db/models/factories.py:61]: @chouinar you should have access now to the bid repo, try again

File: app/api/route/handler/example_person_handler.py (hunk 8 above)
Comment (chouinar) [reply app/api/route/handler/example_person_handler.py:32]: I created a follow-up issue to investigate if there's a way to make the validation more straightforward (not a weird mix of openapi+pydantic): https://github.com/navapbc/template-application-flask/issues/6 along with a few potential solutions.

**Comment by chouinar** [comment]
> 

I meant risky in terms of writing a migration that is wrong. The main benefit of the auto-generated migrations from alembic that exist right now is that (aside from a small set of edge cases) they just work.

File: app/api/db/models/example_models.py (hunk 6 above)
Comment (chouinar) [reply app/api/db/models/example_models.py]: Happy to adjust the example (maybe something like `ExampleRole`?), but I do think we'd need to add some pretty thorough documentation explaining the organizational philosophy of the models, otherwise a developer will just see "oh hey, you can put the models in multiple files" and they'll just organize arbitrarily.

I'm still not sure if we're in agreement or not on the right way to structure this as reading through the documents you linked above isn't quite clicking with me on the distinction you're trying to make.

File: app/api/db/models/factories.py (hunk 5 above)
Comment (chouinar) [reply app/api/db/models/factories.py:61]: I think there's two aspects to data generation at play. General data generation, and populating a test DB.

I do like the approach you linked regarding generating data regardless of the context. Might be worth a follow-up to add something similar.

What would be your suggestion for populating a DB with test scenarios? Assuming we use an ORM, the factory approach does help with setting up several tables at once via the relationships. Without the ORM though, I'm curious how you would structure that? Ideally not needing to manually insert all the records into the DB (which I've done on projects before which just leads to developers not testing the complex cases).

Comment (lorenyu) [reply app/api/db/models/factories.py:61]: > What would be your suggestion for populating a DB with test scenarios? Assuming we use an ORM, the factory approach does help with setting up several tables at once via the relationships. Without the ORM though, I'm curious how you would structure that? Ideally not needing to manually insert all the records into the DB (which I've done on projects before which just leads to developers not testing the complex cases).

we might need to dig deeper into some examples to figure this out, since i think there are different testing situations that would merit different approaches.

i think there's the low level testing of the repository/persistence layer itself (no business logic, just testing that the queries work, that low level inserts, updates, and deletes work and don't leave the database in an inconsistent state). for this we need to populate the DB with varying fixtures to test out the query logic, and i agree with you that some sort of factory pattern could work well for that.

for testing out the more complex services layer that has the business logic, that topic probably merits more discussion with some examples.

File: app/api/db/models/example_models.py (hunk 6 above)
Comment (lorenyu) [reply app/api/db/models/example_models.py]: yeah, I think `User` and `Role` would be great examples, with `Role` and `User` both defined in the `models/user.py` module.

good point about needing documentation regardless

File: app/api/db/migrations/versions/2022_09_12_16_18_41_64cca9571cf0_first_migration.py (hunk 2 above)
Comment (lorenyu) [reply app/api/db/migrations/versions/2022_09_12_16_18_41_64cca9571cf0_first_migration.py:22]: i guess i was thinking that if the developer wanted to add a user table they could just use the example one. then that's one less thing they have to do. it feels like a bonus feature if the template has something you need already. for example, i think one of the template API routes could be a `/health` healthcheck route that queries `SELECT 1` from the database to check the database connection.

**Comment by lorenyu** [comment]
> I meant risky in terms of writing a migration that is wrong.

Yeah that's a good point. I could definitely be convinced

File: app/api/db/migrations/versions/2022_09_12_16_18_41_64cca9571cf0_first_migration.py (hunk 2 above)
Comment (chouinar) [reply app/api/db/migrations/versions/2022_09_12_16_18_41_64cca9571cf0_first_migration.py:22]: There is already a healthcheck endpoint that does that, just not modified in this PR to be visible: https://github.com/navapbc/wic-mt-demo-project-mock-api/blob/main/app/api/route/healthcheck.py

File: app/api/db/models/example_models.py (hunk 6 above)
Comment (chouinar) [reply app/api/db/models/example_models.py]: I might punt switching to user/role to the ticket where I add more endpoints as we'd need more than one endpoint to have that API make sense (at least a `create role` and `create user` one) and I'd like to keep this contained to purely "making it generic".

Unless you've got any other ideas that might work with the shape I've already created for this endpoint. Was aiming for a "core" record with a list of associated records (in this case, a person has multiple pets) that logically could be done in a single POST.

Comment (lorenyu) [reply app/api/db/models/example_models.py]: sounds good to me

File: app/api/db/models/factories.py (hunk 5 above)
Comment (chouinar) [reply app/api/db/models/factories.py:61]: Added https://github.com/navapbc/template-application-flask/issues/7 as a general DB/data follow-up

**Comment by chouinar** [comment]
I think I've accounted for most of the feedback (or created a follow-up ticket for further investigation).

The two primary follow-ups I would plan:
* Adjust the endpoints a bit to be more realistic use cases (User and Role (maybe ?)) - along with this add a few additional endpoint types like patch/get.
* Adding additional documentation.

**Review by rocketnova — APPROVED** [review]
LGTM!
